### PR TITLE
Add support for Spanner BatchCreateSessions

### PR DESCRIPTION
**Pull request by Jennnnny**
The algorithm of picking a new channel has changed, too
It will create a new channel first if the number of channels is not maxed out.

### CODE REVIEW

File: grpc-gcp/src/main/java/com/google/grpc/gcp/GcpManagedChannel.java [hunk 1]
```diff
@@ -356,14 +372,15 @@ static String getKeyFromMessage(MessageOrBuilder msg, String name) {
...
-      String key = getKeyFromMessage((MessageOrBuilder) message, keyName);
-      if (key != null && key != "") {
-        if ((isReq && (cmd == AffinityConfig.Command.UNBIND || cmd == AffinityConfig.Command.BOUND))
-            || (!isReq && cmd == AffinityConfig.Command.BIND)) {
-          {
-            return key;
-          }
+      List<String> keys = getKeyFromMessage((MessageOrBuilder) message, keyName);
+      if (isReq && (cmd == AffinityConfig.Command.UNBIND || cmd == AffinityConfig.Command.BOUND)) {
+        if (keys.size() > 1) {
```
Comment (WeiranFang) [inline grpc-gcp/src/main/java/com/google/grpc/gcp/GcpManagedChannel.java]: `keys.size()` can be 0 as well, as I can see from the logic in `getKeyFromMessage`.

Also, can we also support multiple unbindings? Spanner may have BatchDeleteSessions as well in future.

File: grpc-gcp/src/main/java/com/google/grpc/gcp/GcpClientCall.java [hunk 2]
```diff
@@ -181,11 +184,11 @@ public void onClose(Status status, Metadata trailers) {
           delegateChannelRef.activeStreamsCountDecr();
         }
         // If the operation completed successfully, bind/unbind the affinity key.
-        if (key != null && status.getCode() == Status.Code.OK) {
+        if (keys != null && status.getCode() == Status.Code.OK) {
           if (affinity.getCommand() == AffinityConfig.Command.UNBIND) {
-            delegateChannel.unbind(key);
+            delegateChannel.unbind(keys.get(0));
```
Comment (WeiranFang) [inline grpc-gcp/src/main/java/com/google/grpc/gcp/GcpClientCall.java]: I think `keys` is not guaranteed to have size > 0

File: grpc-gcp/src/main/java/com/google/grpc/gcp/GcpManagedChannel.java (hunk 1 above)
Comment (Jennnnny) [reply grpc-gcp/src/main/java/com/google/grpc/gcp/GcpManagedChannel.java]: Done

File: grpc-gcp/src/main/java/com/google/grpc/gcp/GcpManagedChannel.java [hunk 3]
```diff
@@ -346,7 +362,7 @@ static String getKeyFromMessage(MessageOrBuilder msg, String name) {
    * @param message the <reqT> or <respT> prototype message.
    * @param isReq indicates if the message is a request message.
    */
-  protected <ReqT, RespT> String checkKey(
+  protected <ReqT, RespT> List<String> checkKey(
```
[the remark this answers is not in the document]
Comment (Jennnnny) [reply grpc-gcp/src/main/java/com/google/grpc/gcp/GcpManagedChannel.java]: Done

File: grpc-gcp/src/main/java/com/google/grpc/gcp/GcpManagedChannel.java [hunk 4]
```diff
@@ -318,26 +320,40 @@ private void loadApiConfig(ApiConfig apiConfig) {
    * "session1.session2".
    */
   @VisibleForTesting
-  static String getKeyFromMessage(MessageOrBuilder msg, String name) {
+  static List<String> getKeyFromMessage(MessageOrBuilder msg, String name) {
```
[the remark this answers is not in the document]
Comment (Jennnnny) [reply grpc-gcp/src/main/java/com/google/grpc/gcp/GcpManagedChannel.java]: Done

**Pull request merged**